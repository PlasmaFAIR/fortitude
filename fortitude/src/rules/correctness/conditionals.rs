use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::{find_newline, LineEnding, SourceFile};
use ruff_text_size::TextSize;
use tree_sitter::Node;

/// ## What does it do?
/// Checks for misleading semicolons in `if` statements.
///
/// ## Why is this bad?
/// The following code may appear to execute two statements only if the `if`
/// condition is true, but in actuality the second statement will always be
/// executed:
///
/// ```f90
/// if (condition) print *, "Hello"; print *, "World"
/// ```
///
/// It is equivalent to:
///
/// ```f90
/// if (condition) print *, "Hello"
/// print *, "World"
/// ```
///
/// Users should be cautious applying this fix. If the intent was to have
/// both statements execute only if the condition is true, then the user
/// should rewrite the code to use an `if` statement with a block:
///
/// ```f90
/// if (condition) then
///     print *, "Hello"
///     print *, "World"
/// end if
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MisleadingInlineIfSemicolon {}

impl AlwaysFixableViolation for MisleadingInlineIfSemicolon {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Semicolon following inline if-statement is misleading".to_string()
    }

    fn fix_title(&self) -> String {
        "Replace with newline, or convert to `if(...) then` statement".to_string()
    }
}
impl AstRule for MisleadingInlineIfSemicolon {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // If this is an `if (...) then` construct, exit early
        if !inline_if_statement(node) {
            return None;
        }

        // Check that the if statement is followed directly by a semicolon
        if let Some(semicolon_node) = node.next_sibling().filter(|n| n.kind() == ";") {
            // Find the first following non-semicolon node, and check that it's
            // on the same line as the first semicolon node.
            // Comments on the same line are permitted.
            let mut current_node = semicolon_node;
            while let Some(next) = current_node.next_sibling() {
                if next.kind() == ";" {
                    current_node = next;
                    continue;
                }
                if next.start_position().row != semicolon_node.start_position().row
                    || next.kind() == "comment"
                {
                    return None;
                }
                break;
            }
            // Replace semicolon node with a newline to make it clear that the
            // second statement is not part of the if statement. We should also
            // take care to handle indentation and other whitespace.
            let start = semicolon_node.start_textsize();
            let mut end = semicolon_node.end_byte();
            let text = src.source_text().as_bytes();
            while text[end] == b' ' || text[end] == b'\t' {
                end += 1;
            }
            let end = TextSize::try_from(end).unwrap();
            let indentation = node.indentation(src);
            let edit = Edit::replacement(format!("\n{indentation}"), start, end);
            return some_vec!(Diagnostic::from_node(
                MisleadingInlineIfSemicolon {},
                &semicolon_node
            )
            .with_fix(Fix::safe_edit(edit)));
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["if_statement"]
    }
}

/// ## What does it do?
/// Checks for misleading line continuations in inline `if` statements.
///
/// ## Why is this bad?
/// An inline `if` statement followed immediately by a line continuation
/// can be easily confused for a block `if` statement:
///
/// ```f90
/// if (condition) &
///     call a_very_long_subroutine_name(with, many, ..., arguments)
/// ```
///
/// If a developer wishes to add a second statement to the `if` 'block',
/// they may be tempted to write:
///
/// ```f90
/// if (condition) &
///     call a_very_long_subroutine_name(with, many, ..., arguments)
///     call another_subroutine(args, ...)  ! Always executes!
/// ```
///
/// To avoid this confusion, inline `if` statements that spill over multiple
/// lines should be written as an if-then-block:
///
/// ```f90
/// if (condition) then
///     call a_very_long_subroutine_name(with, many, ..., arguments)
///     call another_subroutine(args, ...)  ! Only executes if condition is true
/// end if
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct MisleadingInlineIfContinuation {}

impl AlwaysFixableViolation for MisleadingInlineIfContinuation {
    #[derive_message_formats]
    fn message(&self) -> String {
        "Line continuation in inline if-statement is misleading".to_string()
    }

    fn fix_title(&self) -> String {
        "Convert to `if(...) then` block".to_string()
    }
}
impl AstRule for MisleadingInlineIfContinuation {
    fn check(_settings: &Settings, node: &Node, src: &SourceFile) -> Option<Vec<Diagnostic>> {
        // If this is an `if (...) then` construct, exit early
        if !inline_if_statement(node) {
            return None;
        }

        // Check if the condition is immediately following by a continuation
        let body_start = node
            .child_with_name("parenthesized_expression")?
            .next_sibling()?;
        if body_start.kind() == "&" {
            let content = ifthenify(node, src)?;
            let start_byte = node.start_textsize();
            let end_byte = node.end_textsize();
            let edit = Edit::replacement(content, start_byte, end_byte);
            return some_vec!(
                Diagnostic::from_node(MisleadingInlineIfContinuation {}, node)
                    .with_fix(Fix::safe_edit(edit))
            );
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["if_statement"]
    }
}

/// Given an if statement, return true if it is inline.
pub fn inline_if_statement(node: &Node) -> bool {
    node.child_with_name("end_if_statement").is_none()
}

/// Given an if statement node, convert it from a one-line if statement
/// to a block if statement.
/// Returns None if the node is already a block if statement.
pub fn ifthenify(node: &Node, src: &SourceFile) -> Option<String> {
    // check that the node is an if statement without an end if statement
    if !inline_if_statement(node) {
        return None;
    }
    let text = node.to_text(src.source_text())?.trim();
    let bytes = text.as_bytes();
    // If CrLf line endings are used, be careful to keep them in the fix
    let nl = find_newline(text)
        .map(|(_, ending)| ending)
        .unwrap_or(LineEnding::Lf)
        .as_str();
    // Divide the if statement into everything up to the end of the condition
    // and the body.
    let condition_node = node.child_with_name("parenthesized_expression")?;
    let condition_end_byte = condition_node.end_byte() - node.start_byte();
    let mut body_start_byte = condition_end_byte;
    // Skip over any whitespace or line continuations between the condition
    // and the body.
    while body_start_byte < node.end_byte()
        && (bytes[body_start_byte].is_ascii_whitespace() || bytes[body_start_byte] == b'&')
    {
        body_start_byte += 1;
    }
    // Build the new if statement.
    let prelude = &text[..condition_end_byte];
    let body = &text[body_start_byte..];
    let indentation = node.indentation(src);
    // Assume two-space indent for the if body
    Some(format!(
        "{} then{nl}{indentation}  {}{nl}{indentation}end if",
        prelude, body
    ))
}
