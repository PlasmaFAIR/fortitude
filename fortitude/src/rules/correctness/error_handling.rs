use std::iter::once;

use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use anyhow::{anyhow, Context, Result};
use ruff_diagnostics::{Diagnostic, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

#[derive(
    Debug,
    Clone,
    Copy,
    PartialEq,
    Eq,
    strum_macros::Display,
    strum_macros::IntoStaticStr,
    strum_macros::EnumString,
)]
#[strum(serialize_all = "lowercase", ascii_case_insensitive)]
enum StatType {
    Stat,
    IoStat,
    CmdStat,
}

impl StatType {
    fn from_node(node: &Node, src: &str) -> Result<Self> {
        match node.kind() {
            "allocate_statement" | "deallocate_statement" => Ok(StatType::Stat),
            "open_statement"
            | "close_statement"
            | "read_statement"
            | "write_statement"
            | "inquire_statement"
            | "file_position_statement" => Ok(StatType::IoStat),
            "subroutine_call" => {
                // Looking only for execute_command_line
                let subroutine_node = node
                    .child_by_field_name("subroutine")
                    .context("Could not retrieve subroutine name")?;
                let subroutine_text = subroutine_node
                    .to_text(src)
                    .context("Failed to parse subroutine text")?
                    .to_lowercase();
                if subroutine_text == "execute_command_line" {
                    Ok(StatType::CmdStat)
                } else {
                    Err(anyhow!("Unknown subroutine: {subroutine_text}"))
                }
            }
            _ => Err(anyhow!("Node does not have a stat type")),
        }
    }

    fn errmsg(self) -> &'static str {
        match self {
            StatType::Stat => "errmsg",
            StatType::IoStat => "iomsg",
            StatType::CmdStat => "cmdmsg",
        }
    }
}

enum CheckStatus {
    Checked,
    Unchecked,
    Overwritten,
}

/// ## What does it do?
/// This rule detects whether a `stat`, `iostat`, and `cmdstat` argument is checked
/// within the same scope it is set.
///
/// ## Why is this bad?
/// By default, `allocate` statements will abort the program if the allocation
/// fails. This is often the desired behaviour, but to provide for cases in
/// which the developer wants to handle allocation errors gracefully, they may
/// optionally check the status of an `allocate` statement by passing a variable
/// to the `stat` argument:
///
/// ```f90
/// allocate (x(BIG_INT), stat=status)
/// if (status /= 0) then
///   call handle_error(status)
/// end if
/// ```
///
/// However, if the `stat` variable is not checked, the program will continue to
/// run despite the allocation failure, which can lead to undefined behaviour.
/// Similar behaviour is exhibited by `deallocate` and IO statements such as
/// `open`, `read`, and `close`.
///
/// To avoid confusing and bug-prone control flow, the checks on status parameters
/// should occur within the same scope in which they are set.
#[derive(ViolationMetadata)]
pub(crate) struct UncheckedStat {
    name: String,
    stat: StatType,
    result: CheckStatus,
}

impl Violation for UncheckedStat {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name, stat, result } = self;
        let check_status = match result {
            CheckStatus::Checked => "checked",
            CheckStatus::Unchecked => "not checked",
            CheckStatus::Overwritten => "overwritten before being checked",
        };
        format!("{stat} argument '{name}' is {check_status} in this scope.")
    }
}

impl AstRule for UncheckedStat {
    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = source.source_text();

        // Check this is an error checking statement, and get the stat type
        let stat_type = StatType::from_node(node, src).ok()?;
        let stat_name: &'static str = stat_type.into();

        // Find a 'stat' argument in the allocate statement
        let arg_list = if node.kind() == "subroutine_call" {
            node.child_with_name("argument_list")?
        } else {
            *node
        };
        let stat_node = arg_list.kwarg(stat_name, src)?;

        let name = stat_node
            .child_by_field_name("value")?
            .to_text(src)?
            .to_string();

        // Check if the 'stat' variable is checked.
        //
        // - Scan all siblings of the allocate statement and their descendants
        //
        // - Find an instance of the variable being used somewhere.
        //
        // - If we reach the end of the siblings without finding one, try again
        //   from the sibling's ancestors. This is to cover cases where the allocate
        //   statement is nested in something like an if statement, but the variable
        //   is checked in the parent scope, e.g.:
        //
        // ```f90
        // if (twice_as_big) then
        //   allocate (x(2*BIG_INT), stat=status)
        // else
        //   allocate (x(BIG_INT), stat=status)
        // end if
        // if (status /= 0) then
        //   call handle_error(status)
        // end if
        // ```
        // - If we reach the end of the current function, subroutine, program, block,
        //   or module procedure without finding the variable, then we consider it a
        //   violation.
        //
        // - If the first use is in another allocate statement, or the left hand side
        //   of an assignment statement, consider that a violation regardless of other
        //   factors.

        for ancestor in once(*node)
            .chain(node.ancestors())
            .take_while(not_scope_boundary)
        {
            match find_stat_in_siblings(&ancestor, &name.to_lowercase(), src) {
                Ok(CheckStatus::Checked) => {
                    // Found the variable, so stop checking.
                    return None;
                }
                Ok(CheckStatus::Overwritten) => {
                    // Found the variable, but it has been overwritten.
                    return some_vec!(Diagnostic::from_node(
                        Self {
                            name,
                            stat: stat_type,
                            result: CheckStatus::Overwritten
                        },
                        &stat_node
                    ));
                }
                Ok(CheckStatus::Unchecked) => {
                    // Didn't find it here. Continue searching.
                    continue;
                }
                _ => {
                    // Something went wrong, bail out.
                    return None;
                }
            }
        }
        some_vec!(Diagnostic::from_node(
            Self {
                name,
                stat: stat_type,
                result: CheckStatus::Unchecked
            },
            &stat_node
        ))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "allocate_statement",
            "deallocate_statement",
            "open_statement",
            "close_statement",
            "read_statement",
            "write_statement",
            "inquire_statement",
            "file_position_statement",
            "subroutine_call", // for execute_command_line
        ]
    }
}

fn find_stat_in_siblings(node: &Node, stat_name: &str, src: &str) -> Result<CheckStatus> {
    let mut sibling = *node;

    while let Some(node) = sibling.next_sibling() {
        sibling = node;

        if new_branch(&sibling) {
            // Ignore occurences in else, elseif, and case statements,
            // as these are out of reach from the current assignment.
            continue;
        }

        if let Some(stat_node) = once(sibling).chain(sibling.descendants()).find(|d| {
            d.kind() == "identifier"
                && d.to_text(src)
                    .is_some_and(|d| d.to_lowercase().as_str() == stat_name)
        }) {
            return stat_check_status(&stat_node, stat_name, src);
        }
    }
    Ok(CheckStatus::Unchecked)
}

fn stat_check_status(node: &Node, stat_name: &str, src: &str) -> Result<CheckStatus> {
    let ancestor = node.parent().context("Node should have a parent")?;

    // stat_name should be lowercase.
    // Two cases to consider:
    //
    // - The stat variable is on the left hand side of an assignment statement.
    // - The stat variable is passed to another error handling parameter.
    //
    // We expect false negatives if stat is passed to a user function or
    // subroutine and overwritten/ignored there.

    if ancestor.kind() == "assignment_statement" {
        if let Some(lhs) = ancestor.child_by_field_name("left") {
            let lhs_text = lhs.to_text(src).context("to_text error")?;
            if lhs_text.to_lowercase().as_str() == stat_name {
                return Ok(CheckStatus::Overwritten);
            }
        }
    }
    if ancestor.kind() == "keyword_argument" {
        // See if the stat variable is passed to another error handling routine.
        let routine = ancestor
            .parent()
            .context("Keyword argument should have a parent")?;
        // If the parent is an argument list, then the routine is the grandparent.
        let routine = if routine.kind() == "argument_list" {
            routine
                .parent()
                .context("Argument list should have a parent")?
        } else {
            routine
        };
        let is_in_error_checking_routine = StatType::from_node(&routine, src).is_ok();
        let kwarg_name_is_stat_type = StatType::try_from(
            ancestor
                .child_by_field_name("name")
                .context("Keyword argument should have a name")?
                .to_text(src)
                .context("to_text error")?,
        )
        .is_ok();
        if is_in_error_checking_routine && kwarg_name_is_stat_type {
            return Ok(CheckStatus::Overwritten);
        }
    }
    Ok(CheckStatus::Checked)
}

fn not_scope_boundary(node: &Node) -> bool {
    !matches!(
        node.kind(),
        "function" | "subroutine" | "program" | "module_procedure" | "block_construct"
    )
}

fn new_branch(node: &Node) -> bool {
    matches!(
        node.kind(),
        "elseif_clause" | "else_clause" | "case_statement"
    )
}

enum AllocationType {
    Allocate,
    Deallocate,
}
impl AllocationType {
    fn from_node(node: &Node) -> Result<Self> {
        match node.kind() {
            "allocate_statement" => Ok(AllocationType::Allocate),
            "deallocate_statement" => Ok(AllocationType::Deallocate),
            _ => Err(anyhow!("Node is not an allocation type")),
        }
    }
}

/// ## What does it do?
/// This rule detects whether `stat` is used alongside multiple allocations or
/// deallocations.
///
/// ## Why is this bad?
/// When allocating or deallocating multiple variables at once, the use of a `stat`
/// parameter will permit the program to continue running even if one of the
/// allocations or deallocations fails. However, it may not be clear which
/// allocation or deallocation caused the error.
///
/// To avoid confusion, it is recommended to use separate allocate or deallocate
/// statements for each variable and check the `stat` parameters individually.
#[derive(ViolationMetadata)]
pub(crate) struct MultipleAllocationsWithStat {
    alloc_type: AllocationType,
}

impl Violation for MultipleAllocationsWithStat {
    #[derive_message_formats]
    fn message(&self) -> String {
        let allocations = match self.alloc_type {
            AllocationType::Allocate => "allocations",
            AllocationType::Deallocate => "deallocations",
        };
        format!("'stat' parameter used with multiple {allocations}.")
    }
}

impl AstRule for MultipleAllocationsWithStat {
    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = source.source_text();

        // Check this has a stat parameter
        let stat_node = node.kwarg("stat", src)?;

        // Count allocations
        let count = if node.kind() == "allocate_statement" {
            count_allocations(node)
        } else {
            count_deallocations(node)
        };
        if count <= 1 {
            return None;
        }

        let alloc_type = AllocationType::from_node(node).ok()?;

        some_vec!(Diagnostic::from_node(Self { alloc_type }, &stat_node))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["allocate_statement", "deallocate_statement"]
    }
}

fn count_allocations(node: &Node) -> usize {
    node.children_by_field_name("allocation", &mut node.walk())
        .count()
}

fn count_deallocations(node: &Node) -> usize {
    node.named_children(&mut node.walk())
        .filter(|c| c.kind() == "identifier")
        .count()
}

/// ## What does it do?
/// This rule detects whether `stat` is used without also setting `errmsg` when
/// allocating or deallocating. Similarly checks for the use of `iostat` without
/// `iomsg` with IO routines, and `cmdstat` without `cmdmsg` when using
/// `execute_command_line`.
///
/// ## Why is this bad?
/// The error codes returned when using `stat`, `iostat`, or `cmdstat` are not
/// very informative on their own, and are not portable across compilers. It is
/// recommended to always capture the associated error message alongside the
/// error code:
///
/// ```f90
/// real, allocatable :: x(:)
/// integer :: status
/// character(256) :: message ! N.B. Can be allocatable in F2023+
///
/// allocate (x(100), stat=status, errmsg=message)
/// open (unit=10, file="data.txt", iostat=status, iomsg=message)
/// call execute_command_line("ls", cmdstat=status, cmdmsg=message)
/// ```
#[derive(ViolationMetadata)]
pub(crate) struct StatWithoutMessage {
    stat_type: StatType,
}

impl Violation for StatWithoutMessage {
    #[derive_message_formats]
    fn message(&self) -> String {
        let stat: &'static str = self.stat_type.into();
        let errmsg = self.stat_type.errmsg();
        format!("'{stat}' used without '{errmsg}'.")
    }
}

impl AstRule for StatWithoutMessage {
    fn check(_settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = source.source_text();

        let stat_type = StatType::from_node(node, src).ok()?;
        let stat_name: &'static str = stat_type.into();
        let arg_list = if node.kind() == "subroutine_call" {
            node.child_with_name("argument_list")?
        } else {
            *node
        };
        if let Some(kwarg) = arg_list.kwarg(stat_name, src) {
            if !arg_list.kwarg_exists(stat_type.errmsg(), src) {
                return some_vec!(Diagnostic::from_node(Self { stat_type }, &kwarg));
            }
        }
        None
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "allocate_statement",
            "deallocate_statement",
            "open_statement",
            "close_statement",
            "read_statement",
            "write_statement",
            "inquire_statement",
            "file_position_statement",
            "subroutine_call", // for execute_command_line
        ]
    }
}
