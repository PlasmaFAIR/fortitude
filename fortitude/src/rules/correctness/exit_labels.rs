use crate::ast::FortitudeNode;
use crate::settings::Settings;
use crate::{AstRule, FromAstNode};
use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix, FixAvailability, Violation};
use ruff_macros::{derive_message_formats, ViolationMetadata};
use ruff_source_file::SourceFile;
use tree_sitter::Node;

/// ## What does it do?
/// When using `exit` or `cycle` in a named `do` loop, the `exit`/`cycle` statement
/// should use the loop name
///
/// ## Example
/// ```f90
/// name: do
///   exit name
/// end do name
/// ```
///
/// Using named loops is particularly useful for nested or complicated loops, as it
/// helps the reader keep track of the flow of logic. It's also the only way to `exit`
/// or `cycle` outer loops from within inner ones.
#[derive(ViolationMetadata)]
pub(crate) struct MissingExitOrCycleLabel {
    name: String,
    label: String,
}

impl Violation for MissingExitOrCycleLabel {
    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;

    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name, label } = self;
        format!("'{name}' statement in named 'do' loop missing label '{label}'")
    }

    fn fix_title(&self) -> Option<String> {
        let Self { label, .. } = self;
        Some(format!("Add label '{label}'"))
    }
}
impl AstRule for MissingExitOrCycleLabel {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // Skip unlabelled loops
        let label = node
            .child_with_name("block_label_start_expression")?
            .to_text(src)?
            .trim_end_matches(':');

        let violations: Vec<Diagnostic> = node
            .named_descendants_except(["do_loop_statement"])
            .filter(|node| node.kind() == "keyword_statement")
            .map(|stmt| (stmt, stmt.to_text(src).unwrap_or_default().to_lowercase()))
            .filter(|(_, name)| name == "exit" || name == "cycle")
            .map(|(stmt, name)| {
                let label_with_space = format!(" {label}");
                let edit = Edit::insertion(label_with_space, stmt.end_textsize());
                let fix = Fix::unsafe_edit(edit);

                Diagnostic::from_node(
                    Self {
                        name: name.to_string(),
                        label: label.to_string(),
                    },
                    &stmt,
                )
                .with_fix(fix)
            })
            .collect();

        Some(violations)
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["do_loop_statement"]
    }
}

/// ## What does it do?
/// Checks for `exit` or `cycle` in unnamed `do` loops
///
/// ## Why is this bad?
/// Using loop labels with `exit` and `cycle` statements prevents bugs when exiting the
/// wrong loop, and helps readability in deeply nested or long loops. The danger is
/// particularly enhanced when code is refactored to add further loops.
///
/// ## Settings
/// See [allow-unnested-loops](../settings.md#check_exit-unlabelled-loops_allow-unnested-loops)
#[derive(ViolationMetadata)]
pub(crate) struct ExitOrCycleInUnlabelledLoop {
    name: String,
}

impl Violation for ExitOrCycleInUnlabelledLoop {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self { name } = self;
        format!("'{name}' statement in unlabelled 'do' loop")
    }
}

impl AstRule for ExitOrCycleInUnlabelledLoop {
    fn check(settings: &Settings, node: &Node, source: &SourceFile) -> Option<Vec<Diagnostic>> {
        let src = source.source_text();
        let name = node.to_text(src)?.to_lowercase();
        // This filters to the keywords we want that _also_ don't have a label
        if !matches!(name.as_str(), "exit" | "cycle") {
            return None;
        }

        let parent_loop = node
            .ancestors()
            .filter(|ancestor| ancestor.kind() == "do_loop_statement")
            .nth(0)?;

        // Immediate parent loop has a label, but we don't want to warn here, because
        // that's covered by missing-exit-or-cycle-label
        if parent_loop
            .child_with_name("block_label_start_expression")
            .is_some()
        {
            return None;
        }

        // If we're only supposed to check on nested loops, check that there is at least
        // one more level of nesting
        if settings.check.exit_unlabelled_loops.allow_unnested_loops {
            parent_loop
                .ancestors()
                .filter(|ancestor| ancestor.kind() == "do_loop_statement")
                .nth(0)?;
        }

        some_vec!(Diagnostic::from_node(Self { name }, node))
    }

    fn entrypoints() -> Vec<&'static str> {
        vec!["keyword_statement"]
    }
}

/// ## What does it do?
/// Check for missing block labels on `end` statements
///
/// ## Why is this bad?
/// This is a compile error!
#[derive(ViolationMetadata)]
pub(crate) struct MissingEndLabel {
    start_name: String,
    end_name: String,
    label: String,
}

impl AlwaysFixableViolation for MissingEndLabel {
    #[derive_message_formats]
    fn message(&self) -> String {
        let Self {
            start_name,
            end_name,
            ..
        } = self;
        format!("'{end_name}' statement in named '{start_name}' block missing label")
    }

    fn fix_title(&self) -> String {
        let Self { label, .. } = self;
        format!("Add label '{label}'")
    }
}

impl AstRule for MissingEndLabel {
    fn check<'a>(
        _settings: &Settings,
        node: &'a Node,
        src: &'a SourceFile,
    ) -> Option<Vec<Diagnostic>> {
        let src = src.source_text();
        // Skip unlabelled loops
        let label = node
            .child_with_name("block_label_start_expression")?
            .to_text(src)?
            .trim_end_matches(':');

        let end = match node.kind() {
            "select_case_statement" | "select_type_statement" | "select_rank_statement" => {
                "end_select_statement".to_string()
            }
            _ => format!("end_{}", node.kind()),
        };

        let violation: Diagnostic = node
            .child_with_name(&end)
            .filter(|node| node.child_with_name("block_label").is_none())
            .map(|stmt| (stmt, stmt.to_text(src).unwrap_or_default().to_lowercase()))
            .map(|(stmt, end_name)| {
                let label_with_space = format!(" {label}");
                let edit = Edit::insertion(label_with_space, stmt.end_textsize());
                let fix = Fix::safe_edit(edit);

                let start_name = node.child(1).unwrap().to_text(src).unwrap().to_lowercase();

                Diagnostic::from_node(
                    Self {
                        end_name: end_name.to_string(),
                        start_name,
                        label: label.to_string(),
                    },
                    &stmt,
                )
                .with_fix(fix)
            })?;

        Some(vec![violation])
    }

    fn entrypoints() -> Vec<&'static str> {
        vec![
            "associate_statement",
            "block_construct",
            "coarray_critical_statement",
            "coarray_team_statement",
            "do_loop_statement",
            "forall_statement",
            "if_statement",
            "select_case_statement",
            "select_rank_statement",
            "select_type_statement",
            "where_statement",
        ]
    }
}

pub(crate) mod settings {
    use crate::display_settings;
    use ruff_macros::CacheKey;
    use std::fmt::{Display, Formatter};

    #[derive(Debug, Clone, Default, CacheKey)]
    pub struct Settings {
        pub allow_unnested_loops: bool,
    }

    impl Display for Settings {
        fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
            display_settings! {
                formatter = f,
                namespace = "check.exit_unlabelled_loops",
                fields = [self.allow_unnested_loops]
            }
            Ok(())
        }
    }
}
