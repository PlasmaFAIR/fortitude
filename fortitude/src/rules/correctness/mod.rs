pub mod accessibility_statements;
pub mod assumed_size;
pub mod derived_default_init;
pub mod exit_labels;
pub mod external;
pub mod implicit_kinds;
pub mod implicit_typing;
pub mod init_decls;
pub mod intent;
pub mod kind_suffixes;
pub mod magic_numbers;
pub mod missing_io_specifier;
pub mod select_default;
pub mod trailing_backslash;
pub mod use_statements;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::apply_common_filters;
    use crate::registry::Rule;
    use crate::rules::correctness::exit_labels;
    use crate::settings::{CheckSettings, Settings};
    use crate::test::test_path;

    #[test_case(Rule::ImplicitTyping, Path::new("C001.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002.f90"))]
    #[test_case(Rule::ImplicitExternalProcedures, Path::new("C003.f90"))]
    #[test_case(Rule::MissingDefaultCase, Path::new("C011.f90"))]
    #[test_case(Rule::NoRealSuffix, Path::new("C021.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("C022.f90"))]
    #[test_case(Rule::MagicNumberInArraySize, Path::new("C031.f90"))]
    #[test_case(Rule::MagicIoUnit, Path::new("C032.f90"))]
    #[test_case(Rule::MissingActionSpecifier, Path::new("C041.f90"))]
    #[test_case(Rule::TrailingBackslash, Path::new("C051.F90"))]
    #[test_case(Rule::MissingIntent, Path::new("C061.f90"))]
    #[test_case(Rule::AssumedSize, Path::new("C071.f90"))]
    #[test_case(Rule::AssumedSizeCharacterIntent, Path::new("C072.f90"))]
    #[test_case(Rule::InitialisationInDeclaration, Path::new("C081.f90"))]
    #[test_case(Rule::PointerInitialisationInDeclaration, Path::new("C082.f90"))]
    #[test_case(Rule::ExternalProcedure, Path::new("C091.f90"))]
    #[test_case(Rule::ProcedureNotInModule, Path::new("C092.f90"))]
    #[test_case(Rule::MissingDefaultPointerInitalisation, Path::new("C101.f90"))]
    #[test_case(Rule::UseAll, Path::new("C121.f90"))]
    #[test_case(Rule::MissingIntrinsic, Path::new("C122.f90"))]
    #[test_case(Rule::MissingAccessibilityStatement, Path::new("C131.f90"))]
    #[test_case(Rule::DefaultPublicAccessibility, Path::new("C132.f90"))]
    #[test_case(Rule::MissingExitOrCycleLabel, Path::new("C141.f90"))]
    #[test_case(Rule::ExitOrCycleInUnlabelledLoop, Path::new("C142.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }

    #[test_case(Rule::ImplicitTyping, Path::new("C001_ok.f90"))]
    #[test_case(Rule::InterfaceImplicitTyping, Path::new("C002_ok.f90"))]
    fn rules_pass(rule_code: Rule, path: &Path) -> Result<()> {
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert!(
            diagnostics.is_empty(),
            "Test source has no warnings, but some were raised:\n{diagnostics}"
        );
        Ok(())
    }

    #[test]
    fn warn_c142_on_nested_loops_only() -> Result<()> {
        let rule_code = Rule::ExitOrCycleInUnlabelledLoop;
        let path = Path::new("C142.f90");
        let snapshot = format!(
            "{}_{}_nested_loops_only",
            rule_code.as_ref(),
            path.to_string_lossy()
        );
        let default = Settings::default();
        #[allow(clippy::needless_update)]
        let settings = Settings {
            check: CheckSettings {
                exit_unlabelled_loops: exit_labels::settings::Settings {
                    allow_unnested_loops: true,
                },
                ..default.check
            },
            ..default
        };
        let diagnostics = test_path(
            Path::new("correctness").join(path).as_path(),
            &[rule_code],
            &settings,
        )?;
        apply_common_filters!();
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
