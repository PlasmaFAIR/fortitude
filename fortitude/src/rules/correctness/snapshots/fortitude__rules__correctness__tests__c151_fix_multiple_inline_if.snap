---
source: fortitude/src/rules/correctness/mod.rs
expression: fixed
snapshot_kind: text
---
        program test
            implicit none
            integer :: i
            if (i == 1) then
              print *, "foo"
            end if
            if (i == 2) then
              print *, "bar"
            end if
            if (i == 3) print *, "baz"
        end program test
